### ISSUE TITLE
executing customized surrogate gradient in GPU

### ISSUE THREAD

**Issue by gwgknudayanga** [fangwei123456/spikingjelly#280]
Hi,

When i set the device for input tensor 'ẍ́ and network 'net' to cuda(), then the gradient of x is always None. So how can i use the the execute this code in GPU appropriately? thanks,

import torch
import torch.nn as nn
from spikingjelly.activation_based import neuron, surrogate

class ste(torch.autograd.Function):
@staticmethod
def forward(ctx, x, alpha):
if x.requires_grad:
ctx.save_for_backward(x)
ctx.alpha = alpha
return surrogate.heaviside(x)

@staticmethod
def backward(ctx, grad_output):
    print(grad_output)
    x = ctx.saved_tensors[0]
    print(x)
    mask = (x.abs() <= ctx.alpha).to(x)
    return grad_output * mask, x.abs().mean()

class STE(nn.Module):
def init(self, alpha: float = 1.):
super(STE, self).init()
self.alpha = nn.Parameter(torch.as_tensor(alpha))

def forward(self, x):
    return ste.apply(x, self.alpha)

net = neuron.IFNode(surrogate_function=STE()).cuda()

x = torch.rand([4, 8], requires_grad=True).cuda()

net(x).sum().backward()
print(x.grad)

**Comment by fangwei123456** [comment]
`then the gradient of x is always None`

https://pytorch.org/docs/stable/notes/autograd.html

> During the backward pass (.backward()), only leaf tensors with requires_grad=True will have gradients accumulated into their .grad fields.

When you use `x = torch.rand([4, 8], requires_grad=True).cuda()`, you can regard that you run the following codes:

```
x_ = torch.rand([4, 8], requires_grad=True)
x = x_.cuda()
```

Only `x_` is the leaf tensor. Thus, when you print `x.grad` you will always get `None`.

You should use the following codes:

```
x = torch.rand([4, 8], device='cuda', requires_grad=True)
```

**Comment by gwgknudayanga** [comment]
Thanks Fangi. this works. However in this case setting the input in cuda, and not setting LIFNode and internal variables of the Surrgoate function (for example 'x' and mask in the following backward function within the Surrogate class) , will that be a issue? (because some tensors are in cuda and other are not specified for cuda).[whatever running the code in that manner didn't give any error.] 

class ste(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, alpha):
        if x.requires_grad:
            ctx.save_for_backward(x)
            ctx.alpha = alpha
        m = surrogate.heaviside(x)
        return m

    @staticmethod
    def backward(ctx, grad_output):
        x = ctx.saved_tensors[0]
        print("saved input tensors ",x)
        mask = (x.abs() <= ctx.alpha).to(x)
        print("grad input from previous ",grad_output)
        grad = grad_output * 0.3 * F.threshold(1.0 - torch.abs(x), 0, 0)
        print("gradient output from this tensor ",grad)
        return grad, None

**Comment by fangwei123456** [comment]
> will that be a issue?

If the neuron and the surrogate do not contain any `tensor`, it will not cause the error. Otherwise pytorch will raise an error.

> for example 'x' and mask in the following backward function within the Surrogate class

`mask` is created from `x` and have the same device with `x`.

**Comment by gwgknudayanga** [comment]
thanks fang, Could you please clarify about the followings 2 things also?

1. In the following at the second time step i am giving a large value (with added 100) as the inputs to the neurons. But when i queried the neuron voltages at the beginning of  third time step, some of the neurons are set to zero voltage ( reset) while many other neurons are not set to zero voltage. Since the threshold is 1 , with this much large input, voltage of all neurons should be set to reset, isn't it?

class VGGSNN(nn.Module):
    def __init__(self):
        super(VGGSNN,self).__init__()
        bias = False
        Affine_flag = True
        self.conv2 = nn.Conv2d(2, 128, kernel_size=3, stride=1, padding=1, bias=bias)
        self.bntt2 = nn.ModuleList([nn.BatchNorm2d(128, eps=1e-4, momentum=0.1, affine=Affine_flag) for i in range(5)])
        self.lif2 = neuron.IFNode()
        self.pool2 = nn.MaxPool2d(kernel_size=2,stride=2)

        self.conv_list = [self.conv2]
        self.bntt_list = [self.bntt2]
        self.lif_list = [self.lif2]


    def forward(self,x):
        
        #print("aaaa ",self.lif_list[0].v)
        for t in range(5):
            print("voltages begining of time step ",self.lif2.v)
            y1 = self.conv_list[0](x[t])
            if t == 2:
                y1 = y1 + 1000
            y2 = self.bntt_list[0][0](y1)
            y3 = self.lif_list[0](y2)
            
        return y3

x = torch.rand([5,2,2,9,9])
myModel = VGGSNN()
myModel(x)


2. I am using ParametricLIFNode Neurons in my network. After initial training of the network i want to set some of those neurons to non trainable while only some others to trainable. So how can i do this? thank you !

**Comment by fangwei123456** [comment]
1. `1000` is not directly sent to the neuron. You need to add `1000` on the output of bn, rather than conv.
2. In PLIF, tau is shared inside a layer. You can only set a whole spiking neurons layer as trainable or not.

> So how can i do this?

https://pytorch.org/docs/stable/notes/autograd.html#setting-requires-grad

**Comment by gwgknudayanga** [comment]
> 1. `1000` is not directly sent to the neuron. You need to add `1000` on the output of bn, rather than conv.
> 
>     2. In PLIF, tau is shared inside a layer. You can only set a whole spiking neurons layer as trainable or not.
> 
> 
> > So how can i do this?
> 
> https://pytorch.org/docs/stable/notes/autograd.html#setting-requires-grad

Thanks Fang !!!. Sof for the second case let me do as follows and proceed so that the tau of this neurons will not be updated and used the preloaded value as it is .
self.lif2 = neuron.IFNode()
self.lif2.requires_grad_ = False;

**Comment by fangwei123456** [comment]
`IFNode` does not contains any learnable parameters.

To make some `PLIFNode` not trainable, refer to this:

https://stackoverflow.com/questions/62523912/how-to-freeze-selected-layers-of-a-model-in-pytorch

**Comment by gwgknudayanga** [comment]
Thanks fang. So is the following way correct? (the following code doesn't give error. But when suggesting parameters , the editor suggests required_grad_ which ends with underscore). Before coding the whole network i need to know this exactly, because why i am asking many. sorry for any inconvenience.

self.conv2 = nn.Conv2d(2, 128, kernel_size=3, stride=1, padding=1, bias=bias)
self.bntt2 = nn.ModuleList([nn.BatchNorm2d(128, eps=1e-4, momentum=0.1, affine=Affine_flag) for i in range(5)])
self.lif2 = neuron.ParametricLIFNode()
self.lif2.requires_grad = False

**Comment by fangwei123456** [comment]
```
for param in self.lif2.parameters():
    param.requires_grad = True
```

**Comment by gwgknudayanga** [comment]
> 
Thanks you Fang !
